### PR TITLE
Added sharing of message types. Fixes #271

### LINKED ISSUE #271: Sharing message types with other roles
Users should be able to share message types with other seats

### PR DESCRIPTION
**Pull request by melvinlu**
This commit draws a message type palette to the left of the room for users to drag-and-drop message types to share among other roles in the room.

# Changes
* Displays a fixed message palette built with a scroll frame to the left of the room graphic
* Dynamic updates (whenever you make/delete/receive a message type, it refreshes)
  * NOTE: the current way of 'refreshing' the palette involves simulating a click of the 'room tab'. There should probably be a better way...however, without refreshing the palette, the focus on the room tab will be messed up on update of the palette (shifts to bottom right)
* **Drag and drop a message type from the palette to a role to share it**
* **Right click a message type from the palette to send it with a pop up dialog**
* **Left click a message type from the palette to view its fields**
* You are notified when...
  * you try and send a message type to yourself
  * successfully send a message type to another role
  * successfully queue up a message type to another role
  * receive a message type from another role
  * receive a message type that you already own from another role
* You can also drag and drop the `when I receive message` and `send message` blocks with their message field set to whatever message type you wish to send

### CODE REVIEW

**Comment by brollb** [comment]
Could you add a message if the user tries to send a message type to a non-existent role? sharemsg

File: src/client/objects.js [hunk 1]
```diff
@@ -958,6 +958,23 @@ SpriteMorph.prototype.deleteMessageType = function(name) {
...
+
+    // Refresh message palette if possible in case the user is already on the 'Room' tab
+    try {
+        this.parentThatIsA(NetsBloxMorph).room.parentThatIsA(ProjectsMorph).updateRoom();
+        // We need to force-update & refresh to fix the layout after drawing the message palette
+        // We can do this by "clicking" the room tab
+        // FIXME: find a better way to refresh...
+        for (var i = 0; i < this.parentThatIsA(NetsBloxMorph).children.length; i++) {
+            if (this.parentThatIsA(NetsBloxMorph).children[i] instanceof Morph) {
+                if (this.parentThatIsA(NetsBloxMorph).children[i].tabBar) {  // found the tab morph
```
Comment (brollb) [inline src/client/objects.js]: no need to iterate here. try:
```
var ide = this.parentThatIsA(NetsBloxMorph);
ide.spriteBar.tabBar
```

File: src/client/room.js [hunk 2]
```diff
@@ -103,6 +105,17 @@ RoomMorph.prototype.update = function(ownerId, name, roles) {
         this.version = Date.now();
         this.drawNew();
         this.changed();
+        // We need to force-update & refresh to fix the layout after drawing the message palette
+        // We can do this by "clicking" the room tab
+        // FIXME: find a better way to refresh...
+        for (var i = 0; i < this.ide.children.length; i++) {
```
Comment (brollb) [inline src/client/room.js]: same as above

File: src/client/room.js [hunk 3]
```diff
@@ -774,9 +903,10 @@ ProjectsMorph.prototype.updateRoom = function() {
     this.addBack(this.contents);
 
     // Draw the room
-    //this.room.setExtent
+    // this.room.setExtent
```
Comment (brollb) [inline src/client/room.js]: remove this

**Comment by brollb** [comment]
The width of the container should also resize to fit the largest message
badwidth

It also might be nice to set the darkened palette holding the message type blocks to match the background since we don't have any border or anything and it looks a little rough as is.

**Comment by melvinlu** [comment]
Thanks for the feedback. I took care of the changes, and in the process of changing the width of the palette to be dynamic, it now only appears when there is a message type to send, which I think looks pretty good.

One thing--I didn't know if you were testing these PRs on the most up-to-date branch or the branch that I derived from, so I left the `NetsBloxSocket.js` message handler for `'share-msg-type'` as both `this.send` AND `this.sendToEveryone` in order for the message to be sent to every client. In build, since you already merged the updated role destinations, it could simply be `this.sendToEveryone`.

**Comment by brollb** [comment]
Looks much better! As a minor thing, could you replace the sent message with a notification (like when you import blocks)?

msgsentdialog

**Comment by melvinlu** [comment]
Yep. Did you just want those two notifications changed? (There are a couple others)

**Comment by brollb** [comment]
In general, the user should not be prompted about stuff unless it is important (especially so if they need to click an 'ok' button - the little notifications can disappear whenever the user gets back to working on stuff). The confirmation dialog can also probably be changed to just a notification - if the recipient doesn't want it, he/she can just deny it.

**Comment by melvinlu** [comment]
What do you mean by deny it? Won't he/she already have received it at that point? So if I understand you correctly, you want the remaining notifications to be changed?
* Attempting to send a message to yourself
* Sending a message to an unoccupied role
* Receive a message you already own

**Comment by brollb** [comment]
yeah, I think those look good. By "deny it", I meant the prompt that the
recipient receives after someone tries to send a message type to him/her
(basically the first message the recipient sees).

**Comment by melvinlu** [comment]
Sorry--'yeah I think those look good' as in you _do_ want those three notifications to be changed or _don't_?

**Comment by brollb** [comment]
I do want them changed

**Comment by melvinlu** [comment]
Ah, gotcha. How's that

**Pull request merged**